components/sdlutil/sdlcursormanager.cpp

```diff
@@ -25,133 +27,57 @@ USE_GRAPHICSWINDOW()
+                for (int r=0; r<source->r(); ++r)
+                    decompressedImage->setColor(source->getColor(s,t,r), s,t,r);
+
+        Uint32 redMask = 0x000000ff;
+        Uint32 greenMask = 0x0000ff00;
+        Uint32 blueMask = 0x00ff0000;
+        Uint32 alphaMask = useAlpha ? 0xff000000 : 0;
+
+        SDL_Surface *cursorSurface = SDL_CreateRGBSurfaceFrom(decompressedImage->data(),
+            width,
+            height,
+            decompressedImage->getPixelSizeInBits(),
+            decompressedImage->getRowSizeInBytes(),
+            redMask,
+            greenMask,
+            blueMask,
+            alphaMask);
+
+        SDL_Surface *targetSurface = SDL_CreateRGBSurface(0, width, height, 32, redMask, greenMask, blueMask, alphaMask);
+        SDL_Renderer *renderer = SDL_CreateSoftwareRenderer(targetSurface);
+
+        SDL_RenderClear(renderer);
+
+        SDL_SetHint(SDL_HINT_RENDER_SCALE_QUALITY, "1");
+        SDL_Texture *cursorTexture = SDL_CreateTextureFromSurface(renderer, cursorSurface);
+
+        SDL_RenderCopyEx(renderer, cursorTexture, NULL, NULL, rotDegrees, NULL, SDL_FLIP_NONE);
+
+        SDL_DestroyTexture(cursorTexture);
+        SDL_FreeSurface(cursorSurface);
+        SDL_DestroyRenderer(renderer);
 
         osg::ref_ptr<osg::Image> resultImage = new osg::Image;
-        resultImage->allocateImage(width, height, 1, GL_RGBA, GL_UNSIGNED_BYTE);
-
-        osg::RenderInfo renderInfo;
-        renderInfo.setState(state);
-
-        glViewport(0, 0, width, height);
-
-        osg::ref_ptr<osg::Geometry> geom;
-
-#if defined(__APPLE__)
-        // Extra flip needed on OS X systems due to a driver bug
-        const char* envval = getenv("OPENMW_CURSOR_WORKAROUND");
-        bool workaround = !envval || envval == std::string("1");
-        std::string vendorString = (const char*)glGetString(GL_VENDOR);
-        if (!envval)
-            workaround = vendorString.find("Intel") != std::string::npos || vendorString.find("ATI") != std::string::npos || vendorString.find("AMD") != std::string::npos;
-        if (workaround)
-            geom = osg::createTexturedQuadGeometry(osg::Vec3(-1,1,0), osg::Vec3(2,0,0), osg::Vec3(0,-2,0));
-        else
-#endif
-        geom = osg::createTexturedQuadGeometry(osg::Vec3(-1,-1,0), osg::Vec3(2,0,0), osg::Vec3(0,2,0));
-
-        geom->drawImplementation(renderInfo);
-
-        glReadPixels(0, 0, width, height, GL_RGBA, GL_UNSIGNED_BYTE, resultImage->data());
-
-        geom->releaseGLObjects();
-        source->releaseGLObjects();
-        texture->releaseGLObjects();
+        resultImage->setFileName(source->getFileName());
+        resultImage->allocateImage(width, height, source->r(), GL_RGBA, GL_UNSIGNED_BYTE);
+        assert(resultImage->getImageSizeInBytes() == (targetSurface->pitch * targetSurface->h));
+        memcpy(resultImage->data(), targetSurface->pixels, resultImage->getImageSizeInBytes());
+        SDL_FreeSurface(targetSurface);
```

Comment:
We may actually return the surface instead of an `osg::Image` to avoid another conversion in `_createCursorFromResource`.